Point cleanup

Allcharles opened the pull request:
# Cleanup point page logic

Finalise and cleanup project/site/point page logic so that the pages work seamlessly with each other. This combines the work of PRs #495 and #488.

PR: #495 

> # Basic Site Belong to a Region Pages

> This PR creates the boilerplate code for implementing the sites belonging to a region pages. It handles setting up forms, menu links, services, etc... The sites belonging to a region have been labelled as "points" so end users can separate them from sites.

PR: #488

> # Implemented basic region pages

> This PR creates the boilerplate code for implementing the region pages. It handles setting up forms, menu links, services, etc...

## Changes

- Added region model
- Added region service
- Created new region page
- Created edit region page
- Created delete region page
- Created region details page
- Modified project details page to display regions
- Modified model permissions widget to handle all model types instead of just project/site
- Added new site belonging to a region page
- Added edit site belonging to a region page
- Added delete site belonging to a region page
- Added site belonging to a region details page
- Created site wizard, this guides the user through creating a site or region depending on their use case
- Enabled `SiteMapComponent` to filter by region
- Created point harvest page

## Problems

- [x] Project details page needs UI changes
- [x] Region details page could use some UI changes
- [x] Region sites currently acts like project sites, these should instead appear as a subgroup of region
- [x] Need to finish implementation of unit tests (await final UI changes)
- [x] Missing the following functionality from regions:
  - [x] Create a site inside a region
  - [x] Edit a site inside a region
  - [x] Delete a site inside a region
  - [x] Most of this relates to point 3
- [x] Haven't fixed any of the review comments related to #488 
- [x] Need to finish implementation of unit tests (awaiting final UI changes)
- [x] Minor bugs related to displaying wrong model name (will fix in final PR)
  - [x] Region Details displays inconsistent filter input placeholder and "found n sites"
  - [x] New, edit, and delete site belonging to region forms have slightly different action menu items
- [x] Edit site belonging to region model fails
- [x] Unit tests need to be updated
- [ ] https://github.com/QutEcoacoustics/workbench-client/pull/495#discussion_r497368412
- [ ] https://github.com/QutEcoacoustics/workbench-client/pull/495#discussion_r497376464

## Issues

Fixes #226 

## Visual Changes

Listed below is the complete list of changes:

### Project Details

#### No Models

![image](https://user-images.githubusercontent.com/3955116/95167580-4621ce00-07f3-11eb-84ba-a0879716dc91.png)

#### With Regions

![image](https://user-images.githubusercontent.com/3955116/95167521-2be7f000-07f3-11eb-8630-339076c9b169.png)

#### With Sites

![image](https://user-images.githubusercontent.com/3955116/95167614-52a62680-07f3-11eb-9d6a-7ca246774f8b.png)

#### With Regions and Sites

![image](https://user-images.githubusercontent.com/3955116/95167553-399d7580-07f3-11eb-92e0-2964ed435502.png)

### New Site

#### Initial Page

![image](https://user-images.githubusercontent.com/3955116/95167742-913be100-07f3-11eb-9b17-346dfa1e8d27.png)

#### New Region

![image](https://user-images.githubusercontent.com/3955116/95169994-18d71f00-07f7-11eb-9900-f8093ff8f6f3.png)

#### New Site

![image](https://user-images.githubusercontent.com/3955116/95167759-9a2cb280-07f3-11eb-8012-7992c427acdc.png)

### Region Details

![image](https://user-images.githubusercontent.com/3955116/95167683-75383f80-07f3-11eb-9de2-4d5e196f23ce.png)

### Edit Region

![image](https://user-images.githubusercontent.com/3955116/95170065-33a99380-07f7-11eb-92fc-55d55e3a87e6.png)

### Delete Region

![image](https://user-images.githubusercontent.com/3955116/95168573-d7de0b00-07f4-11eb-8931-13faf6f86111.png)

### Point Details

![image](https://user-images.githubusercontent.com/3955116/95167894-c34d4300-07f3-11eb-8720-e52bf6f81e00.png)

### New Point

![image](https://user-images.githubusercontent.com/3955116/95169912-fa712380-07f6-11eb-804c-ebaa490509b7.png)

### Edit Point

![image](https://user-images.githubusercontent.com/3955116/95169891-f1805200-07f6-11eb-8905-03bf2c767219.png)

### Delete Point

![image](https://user-images.githubusercontent.com/3955116/95170164-5471e900-07f7-11eb-8e46-b11e22002e57.png)

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [x] Ensure CI build is passing

Allcharles committed:
Created regions service

Allcharles committed:
First look at combined sites and regions

Allcharles committed:
Created basic region pages

Allcharles committed:
Implemented tests for basic forms

Allcharles committed:
Implemented basic regions list

Allcharles committed:
Temporarily disable broken tests

Allcharles committed:
Created PointsModule

Allcharles committed:
Fix minor bugs

Allcharles committed:
Combined regions and sites in details page

Allcharles committed:
Move maps between regions and sites

Allcharles committed:
Cleanup design to match suggestions

Allcharles committed:
Fixed region details page

Allcharles committed:
Cleanup project/region details

Implemented global css variable for highlight colours as part of #503

Allcharles committed:
Extracted point components from site files

Created basic new site wizard component

Allcharles committed:
Implemented wizard

Allcharles committed:
Fixed wizard styling affecting other pages

Allcharles committed:
Final cleanup

Created point harvest page

Allcharles committed:
Updated region json schema

atruskie commented on src/app/components/projects/pages/details/details.component.ts:
```diff
@@ -51,28 +56,109 @@ const projectKey = "project";
+  public regions: List<Region>;
+  public sites: List<Site>;
+  /**
+   * Api requests are independent, use this to track when both complete
+   */
+  private apiReturnCount = 0;
```
requests may be independent, but a hook after the [merge](https://github.com/QutEcoacoustics/workbench-client/pull/511/files#diff-3fc28901b5267a7f8ed292901d944f39R180) could set this to true/false

Allcharles replied:
I'm not really sure how to do this, I can't use `complete` because the parent observable never completes. Alternatively, i could use `last` to read the last value of the merge observable, however that doesn't seem to be possible within the constraints of the abstract parent class

atruskie commented on src/app/components/projects/pages/details/details.component.ts:
```diff
@@ -51,28 +56,109 @@ const projectKey = "project";
+          this.hasRegions = true;
+        }
+
+        this.collectionSize = Math.max(
+          this.collectionSizes.sites,
+          this.collectionSizes.regions
+        );
```
what is this doing?

Allcharles replied:
Trying to find which list of models has the greater maximum page number

Allcharles committed:
Fixed wrong model names

atruskie commented on src/app/components/sites/pages/wizard/wizard.component.ts:
```diff
@@ -0,0 +1,91 @@
+      <p class="lead">Do you have more than one sensor at this site?</p>
+
+      <div class="clearfix">
+        <div class="float-right">
```
I personally prefer to avoid floats at all costs.

Flex can stack right?

atruskie commented on src/styles.scss:
```diff
@@ -5,6 +5,10 @@
 @import "~@swimlane/ngx-datatable/assets/icons.css";
 
+:root {
+  --highlight: #6c757d;
```
Document intention

Also, I'd maybe name it `theme-highlight`

Allcharles committed:
Fix review comments

Allcharles committed:
Fixed site-map tests

Allcharles committed:
Fixed site-card tests

Allcharles committed:
Fixed region and point forms

Allcharles committed:
Fixed site and point detail pages

Allcharles committed:
Upgrade delete pages to use spectator

Allcharles committed:
Fixed region details tests

Allcharles committed:
Fixed project details tests

Fixed minor bugs in project details component:
- Stopped re-render of map component which was causing multiple unneeded requests each time
- Filtered out sites with region ids
- Fixed modelData object not producing html code for faked models

Allcharles committed:
Implement tests for final edge cases

Allcharles committed:
Merge branch 'master' into point-cleanup

atruskie commented on src/app/components/projects/site-card/site-card.component.ts:
```diff
@@ -1,56 +1,77 @@
+  }
+
+  public numPoints() {
+    return this.region.siteIds?.size || 0;
```
```suggestion
    return this.region?.siteIds?.size || 0;
```

Allcharles committed:
Update src/app/components/projects/site-card/site-card.component.ts

Co-authored-by: Anthony Truskinger <anthony.truskinger@gmail.com>
